File: python/299. Bulls and Cows/solution.py
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        bulls = 0
        cows = 0
        secretMap = {}
        guessMap = {}

        for i in range(len(secret)):
            if secret[i] == guess[i]:
                bulls += 1
            else:
                secretMap[secret[i]] = (secretMap.get(secret[i], 0) + 1)
                guessMap[guess[i]]= (guessMap.get(guess[i], 0) + 1)

        for key in guessMap.keys():
            if key in secretMap:
                cows += min(secretMap.get(key), guessMap.get(key))

        return f'{bulls}A{cows}B'

File: java/299. Bulls and Cows/Solution.java
class Solution {
    public String getHint(String secret, String guess) {
        int bulls = 0;
        int cows = 0;
        Map<Character, Integer> secretMap = new HashMap<>();
        Map<Character, Integer> guessMap = new HashMap<>();

        for (int i = 0; i < secret.length(); i++) {
            if (secret.charAt(i) == guess.charAt(i)) {
                bulls++;
            } else {
                secretMap.put(secret.charAt(i), secretMap.getOrDefault(secret.charAt(i), 0) + 1);
                guessMap.put(guess.charAt(i), guessMap.getOrDefault(guess.charAt(i), 0) + 1);
            }
        }

        for (Character c : guessMap.keySet()) {
            if (secretMap.containsKey(c)) {
                cows += Math.min(secretMap.get(c), guessMap.get(c));
            }
        }

        return "" + String.valueOf(bulls) + "A" + String.valueOf(cows) + "B";
    }
}